### PR TITLE
Remote state handling rebase

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

State handling for remote platforms:

Introduce `SimulationState` subclass for remote platform (`RemoteSimulationState` and `PyRemoteSimulationState`) which is essentially a holder of the kernel IR.

Perform a lazy evaluation to resolve the state:

(1) For state accessor, e.g., amplitudes: adding a configurable qubit number threshold. If the number of qubits is less than the threshold, resolve the full state vector to the client. Otherwise, send on amplitude access request. Adding an API to `cudaq::state` for batched amplitude access.

(2) For overlap, sending job requests with both kernel IRs.

Note: the remote state is implemented for MLIR kernels only (C++ MLIR mode, kernel builder and Python).
Library mode C++ kernel is evaluated as-is (v0.7 mode): state is resolved to a vector by the server for each `get_state` call.

Adjust the GPU backend validation: skip a few long running examples for cutensornet-based backends. This brings total GPU validation time to ~2hr.

This is migrated from https://github.com/NVIDIA/cuda-quantum/pull/1653: revise the implementation to make it compatible with pre-existing NVQC services.

### Compatibility

Except for **remote** state overlap and amplitude accessor (large state), it is compatible with existing NVQC service.

### CI 

- Publishing: https://github.com/NVIDIA/cuda-quantum/actions/runs/9476728549

(image: `ghcr.io/nvidia/cuda-quantum:workflow-stateHandling`)

- NVQC regression run: https://github.com/NVIDIA/cuda-quantum/actions/runs/9478140066
(testing against the existing prod NVQC service)

- Nightly integration: https://github.com/NVIDIA/cuda-quantum/actions/runs/9478073356
(deploy `ghcr.io/nvidia/cuda-quantum:workflow-stateHandling` as a test function)

### CODE REVIEW

File: python/runtime/cudaq/algorithms/py_state.cpp
Comment (bettinaheim) [inline python/runtime/cudaq/algorithms/py_state.cpp:436]: Why not go with a string here that is converted to the appropriate integer index/indices during submission? I find this a bit awkward to have to put commas for a large number of qubits (bitvector would also be a natural choice).

**Review by bettinaheim — APPROVED** [review]
Approving to unblock the code ownership, but someone else should take a look at the state changes as well.

File: runtime/common/BaseRestRemoteClient.h
Comment (bmhowe23) [inline runtime/common/BaseRestRemoteClient.h:241]: What is the point of the scope braces on lines 241, 245, 246, and 249?

File: runtime/common/ExecutionContext.h
Comment (bmhowe23) [inline runtime/common/ExecutionContext.h:67]: Just curious ... why is this one non-optional but the other new ones in lines 70 and 74 are optional? If amplitudeMaps is optional as well, I'd recommend making that explicit.

File: runtime/common/JIT.cpp
Comment (bmhowe23) [inline runtime/common/JIT.cpp:73]: nit:
```suggestion
                   << " internal to external linkage.");
```

File: runtime/cudaq/platform/default/rest_server/helpers/RestRemoteServer.cpp
Comment (bmhowe23) [inline runtime/cudaq/platform/default/rest_server/helpers/RestRemoteServer.cpp:618]: It might be helpful to add a reference to `RemoteSimulationState::maxQubitCountForFullStateTransfer` here for anybody searching the code. (It took me a little while to search where "too large" was explicitly defined.)

File: runtime/cudaq/qis/remote_state.cpp
Comment (bmhowe23) [inline runtime/cudaq/qis/remote_state.cpp:133]: Since this is server-side code (I think), what happens if the user tries to submit a request exceeding this limit without providing any specific amplitude basis vectors? Do they get an error message or warning message somewhere?

File: runtime/cudaq/qis/remote_state.h
Comment (1tnguyen) [reply runtime/cudaq/qis/remote_state.h:30]: I added a code comment in https://github.com/NVIDIA/cuda-quantum/pull/1803/commits/8ab176faf836e7ff11a77bd548c1c91f836fecb3 to document this.

This mainly because of the lazy evaluation. The base API from `cudaq::SimulationState` is expected to be `const` but we would perform evaluation at those points. 

The `argsBuffer` data needs to be passed as non-const pointer for those `launchKernel` evaluation calls as we;ll.

File: runtime/cudaq/qis/remote_state.cpp
Comment (1tnguyen) [reply runtime/cudaq/qis/remote_state.cpp:135]: It's actually client-side code. The limit here is mainly an optimization for amplitude access (resolving the whole state vector rather than sending amplitude requests). We'll have a doc update to note about the benefit of using batched amplitudes API rather than sequential amplitude calls for large number of qubits using NVQC.

If users want to resolve the whole state vector (not an amplitude access), we treat them as the current v0.7 mode, sending back the whole state vector (memory allocation failures if the state is too big).

File: runtime/cudaq/platform/default/rest_server/helpers/RestRemoteServer.cpp
Comment (1tnguyen) [reply runtime/cudaq/platform/default/rest_server/helpers/RestRemoteServer.cpp]: Good point. Added in 0549fd7fe1f8f2a653223143028bacdf30839596

File: runtime/common/BaseRestRemoteClient.h
Comment (1tnguyen) [reply runtime/common/BaseRestRemoteClient.h]: I'll defer this change since it'd require further changes to the server-side version check and client-side filter like you said. (more compatibility validation is needed).

**Pull request merged**
